### PR TITLE
Add complex policies (part 1)

### PR DESCRIPTION
**Pull request by KasiaKoz**
Loads of new classes which aim to do everything else they did before this PR plus being able to give more than one probability to a policy and those probabilities can be customised by using custom samplers like in https://github.com/arup-group/pam/pull/16 but returning the p value instead. This enables easier implementation of other modifier policy classes

## Attribute Filter:
- `PersonAttributeFilter` : Handles people/households satisfying conditions, which are defined in the same format as when they were being passed to `RemoveActivity`, instead of `strict_conditions: True/False` we have `how: any/all`

## Policy Level:
- `HouseholdPolicy` : affects all activities in people in household
- `PersonPolicy`: affects all activities in person
- `ActivityPolicy` : affects single activity

## Sampling Level:
- `SamplingProbability` : base class
- `SimpleProbability` : default for single number probabilities given to the Policy level - is applied on the same level as the policy level
   
Below three take either a number or a custom sampler function that returns the `p` value based on input `x, **kwargs`, where `x` is the input at the level of `xProbability` (e.g. `x=household` for `HouseholdProbability`) and `kwargs` are inputs to the custom sampler function (there are examples in [Complex Policy notebook](https://github.com/arup-group/pam/blob/kasia-add-complex-policies/notebooks/PAM_Complex_Policies.ipynb))

- `HouseholdProbability` : computes probability for a household given p number for household
- `PersonProbability` : computes probability for a household or person given p number for person
- `ActivityProbability` : computes probability for a household, person or activity given p number for activity

## Modifier Classes:
- `RemoveActivity` : Now only contains methods to remove activities, sampling/filtering personal attributes is handled by Sampling/AttributeFilter classes respectively
Placeholders for:
- `MoveActivity`
- `AddActivity`

## Usage:
`PolicyLevel(ModifierClass, number or SamplingClass(or list of them), PersonAttributeFilter(optional))`
More examples in [Complex Policy notebook](https://github.com/arup-group/pam/blob/kasia-add-complex-policies/notebooks/PAM_Complex_Policies.ipynb)

Diagram
classes_Pyreverse

### CODE REVIEW

**Comment by elladahan** [comment]
Looks good to me! I was just wondering, were there any tests done during development on bigger datasets to check for any unforeseen errors?

File: pam/parse.py [hunk 1]
```diff
@@ -27,7 +27,7 @@ def load_travel_diary(trips_df, attributes_df, sample_perc = None, complex=True)
 		raise UserWarning("Unrecognised input for population travel diaries")
 
 	if not isinstance(attributes_df, pd.DataFrame):
-	    raise UserWarning("Unrecognised input for population attributes")
+		raise UserWarning("Unrecognised input for population attributes")
```
[the remark this answers is not in the document]
Comment (KasiaKoz) [reply pam/parse.py:30]: lol, I don't know what happened there, and pycharm is showing this to me as the expected indentation in that case (?!) I'm going to run a reformat on this whole file

[the remark this answers is not in the document]
Comment (fredshone) [reply pam/parse.py:30]: It's my fault - i changed editor and did a lot of tab indentation...

**Comment by KasiaKoz** [comment]
> Looks good to me! I was just wondering, were there any tests done during development on bigger datasets to check for any unforeseen errors?

@elladahan nope, we will be running it on a serious dataset over the next couple of days though, all will come out in the wash then 😄

File: tests/test_5_policy_levels.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+def assert_correct_activities(person, ordered_activities_list):
+    assert len(person.plan) % 2 == 1
+    for i in range(0, len(person.plan), 2):
+        assert isinstance(person.plan.day[i], Activity)
+    assert [a.act for a in person.plan.activities] == ordered_activities_list
+    assert person.plan[0].start_time == mtdt(0)
+    assert person.plan[len(person.plan)-1].end_time == END_OF_DAY
+
+
+def test_HouseholdPolicy_verifies_for_appropriate_probabilities(mocker):
```
Comment (mfitz) [inline tests/test_5_policy_levels.py:94]: What is the thing we're most interested in here? That a specific default set of probability types is passed to the helper function when you apply a household policy?

File: tests/test_5_probability_samplers.py [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+
+
+def test_SamplingProbability_samples_when_random_below_prob_val(mocker):
+    mocker.patch.object(modify.SamplingProbability, 'p', return_value=0.55)
+    mocker.patch.object(random, 'random', return_value=0.5)
+    prob = modify.SamplingProbability()
+    assert prob.sample('')
+
+
+def test_SamplingProbability_doesnt_sample_when_random_above_prob_val(mocker):
```
Comment (mfitz) [inline tests/test_5_probability_samplers.py:63]: How about an extra test for when the random value is *equal* to the probability value?

File: tests/test_5_probability_samplers.py [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+def test_HouseholdProbability_accepts_integer():
+    modify.HouseholdProbability(1)
+
+
+def test_HouseholdProbability_fails_non_probability_integers():
+    with pytest.raises(AssertionError):
+        modify.HouseholdProbability(2)
+
+
+def test_HouseholdProbability_accepts_functions():
```
Comment (mfitz) [inline tests/test_5_probability_samplers.py:99]: I'm not completely clear on what we are testing here. Is it that the kwargs passed to the constructor are saved in the probability object?

File: tests/test_5_probability_samplers.py [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+        prob.p(None)
+
+
+def test_PersonProbability_compute_probability_for_household_returns_same_level_p_for_floats():
+    prob = modify.PersonProbability(0.5)
+    assert prob.compute_probability_for_person(Person(1)) == 0.5
+
+
+def test_PersonProbability_compute_probability_for_person_delegates_p_to_custom_callable(mocker):
+    called = None
```
Comment (mfitz) [inline tests/test_5_probability_samplers.py:227]: Glad this works! :-)

File: tests/test_5_probability_samplers.py [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+    custom_sampler('', **prob.kwargs)
+
+
+def test_ActivityProbability_p_delegates_to_compute_probability_for_activity_for_each_activity_for_person_in_Household(mocker, Smith_Household):
+    mocker.patch.object(modify.ActivityProbability, 'compute_probability_for_activity', return_value=0.25)
+    prob = modify.ActivityProbability(['work', 'escort'], 0.25)
+    p = prob.p(Smith_Household)
+
+    assert modify.ActivityProbability.compute_probability_for_activity.call_count == 4
+    assert p == 0.68359375
```
Comment (mfitz) [inline tests/test_5_probability_samplers.py:275]: Why do we expect this value?

File: pam/activity.py [hunk 7]
```diff
@@ -654,6 +654,10 @@ def __str__(self):
 	def __eq__(self, other):
 		return (self.location == other.location) and (self.act == other.act)
 
+	def is_exact(self, other):
+		return (self.location == other.location) and (self.act == other.act) \
```
[the remark this answers is not in the document]
Comment (KasiaKoz) [reply pam/activity.py:658]: yeah it's a tabs indented file, will reformat it

File: tests/test_5_PersonAttributeFilter.py [hunk 8]
```diff
@@ -0,0 +1,149 @@
+from pam import modify
+from pam.core import Household, Person
+from pam.activity import Plan, Activity, Leg
+import pytest
+from pam.utils import minutes_to_datetime as mtdt
+from pam.variables import END_OF_DAY
+from random import choice
+
+
+def instantiate_household_with(persons: list):
```
[the remark this answers is not in the document]
Comment (KasiaKoz) [reply tests/test_5_PersonAttributeFilter.py:10]: yup moving there

File: tests/test_5_MoveActivity.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    Bobby.add(Activity(1, 'home', 'a', start_time=mtdt(0), end_time=mtdt(8 * 60)))
+    Bobby.add(Leg(1, 'walk', 'a', 'b', start_time=mtdt(8 * 60), end_time=mtdt(8 * 60 + 30)))
+    Bobby.add(Activity(2, 'education', 'b', start_time=mtdt(8 * 60 + 30), end_time=mtdt(16 * 60)))
+    Bobby.add(Leg(2, 'walk', 'b', 'c', start_time=mtdt(16 * 60), end_time=mtdt(16 * 60 + 30)))
+    Bobby.add(Activity(5, 'home', 'a', start_time=mtdt(18 * 60 + 30), end_time=END_OF_DAY))
+    return Bobby
+
+
+@pytest.fixture()
+def Smith_Household(Steve, Hilda, Timmy, Bobby):
```
[the remark this answers is not in the document]
Comment (KasiaKoz) [reply tests/test_5_MoveActivity.py:81]: will change

File: tests/test_5_PersonAttributeFilter.py [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+    Bobby.add(Activity(5, 'home', 'a', start_time=mtdt(18 * 60 + 30), end_time=END_OF_DAY))
+    return Bobby
+
+
+@pytest.fixture()
+def Smith_Household(Steve, Hilda, Timmy, Bobby):
+    return instantiate_household_with([Steve, Hilda, Timmy, Bobby])
+
+
+def test_PersonAttributeFilter_satisfies_conditions_delegates_to_household_satisfies_conditions_when_given_household(mocker):
```
[the remark this answers is not in the document]
Comment (KasiaKoz) [reply tests/test_5_PersonAttributeFilter.py:82]: will leave them as they are for now I think as I feel there might be more filters coming

File: tests/test_5_PersonAttributeFilter.py [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+    with pytest.raises(NotImplementedError) as e:
+        modify.PersonAttributeFilter({}).satisfies_conditions('whatever')
+
+
+def test_PersonAttributeFilter_household_satisfies_conditions_returns_True_if_conditions_empty():
+    assert modify.PersonAttributeFilter({}).satisfies_conditions(Household(1))
+
+
+def test_PersonAttributeFilter_household_satisfies_conditions_returns_True_if_one_person_satisfies_conditions(Smith_Household):
+    def condition_age_6(val):
```
[the remark this answers is not in the document]
Comment (KasiaKoz) [reply tests/test_5_PersonAttributeFilter.py:113]: yup, i agree

File: tests/test_5_RemoveActivity.py [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+
+
+def instantiate_household_with(persons: list):
+    household = Household(1)
+    for person in persons:
+        household.add(person)
+    return household
+
+
+@pytest.fixture()
```
[the remark this answers is not in the document]
Comment (KasiaKoz) [reply tests/test_5_RemoveActivity.py:16]: yup, done

File: tests/test_5_policy_levels.py (hunk 2 above)
Comment (KasiaKoz) [reply tests/test_5_policy_levels.py:94]: this checks that the assertion in verify_probability for HouseholdPolicy checks for the right levels of probability that Household can handle i.e. `float, list, modify.SimpleProbability, modify.ActivityProbability, modify.PersonProbability,
         modify.HouseholdProbability`, they vary according to the policy level, it's to do with this [slide](https://docs.google.com/presentation/d/1KjirBAyCfO1aZkAjkF2_f1TFWmq7DhkQ5Xw-BA5zQfI/edit#slide=id.g76ea4112fe_0_45)

File: tests/test_5_probability_samplers.py (hunk 3 above)
Comment (KasiaKoz) [reply tests/test_5_probability_samplers.py:63]: yup

File: tests/test_5_probability_samplers.py (hunk 4 above)
Comment (KasiaKoz) [reply tests/test_5_probability_samplers.py:99]: I was trying to test that `HouseholdProbability` works with a callable, may not have done it very well though, any ideas for improvement?

File: tests/test_5_probability_samplers.py (hunk 6 above)
Comment (KasiaKoz) [reply tests/test_5_probability_samplers.py:275]: 1-(1-0.25)^4

**Pull request merged**